Support decision table inputs and outputs without names/labels

korthout opened the pull request:
## Description

<!-- Please explain the changes you made here. -->
A decision tables' inputs and outputs must have an id and can have a name. What we define as a name differs between inputs and outputs. For inputs, we used only the label from the XML which is not mandatory. For outputs, we used only the name, but it is only mandatory when multiple outputs exist.

Even though the internal decision engine was already able to deal with this, the workflow engine wasn't. It would try to write the EvaluationEvent record with an EvaluatedInputRecord (or EvaluatedOutputRecord) and then fail writing it because of a NullPointerException.

After discussing with the Operate team (who uses this name value), we came to the conclusion that we want the same behavior as the Modeler:

- Inputs can have a label in the XML, when defined we use it as the name. Otherwise, we use the input's expression as name.
- Outputs can have a label in the XML, when defined we use it as the name. Otherwise, we use the output's name.

This PR adjusts which values are used as the names for inputs and outputs of evaluated decision tables. It also makes sure these records can be written when the name of the input or output is undefined.

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #8909

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda-cloud/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [x] I've reviewed my own code
* [x] I've written a clear changelist description
* [x] I've narrowly scoped my changes
* [x] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

korthout commented:
Manually tested to see if it could be successfully exported to ES:

```json
{
  "_index": "zeebe-record_decision-evaluation_1.4.0-snapshot_2022-03-16",
  "_type": "_doc",
  "_id": "1-50",
  "_version": 1,
  "_score": 1,
  "_routing": "1",
  "_source": {
    "partitionId": 1,
    "value": {
      "decisionVersion": 5,
      "decisionOutput": "\"Obi-Wan Kenobi\"",
      "evaluatedDecisions": [
        {
          "decisionType": "DECISION_TABLE",
          "evaluatedInputs": [
            {
              "inputId": "Input_1",
              "inputName": "",
              "inputValue": "\"blue\""
            }
          ],
          "decisionVersion": 4,
          "decisionOutput": "\"Jedi\"",
          "matchedRules": [
            {
              "ruleId": "DecisionRule_0zumznl",
              "ruleIndex": 1,
              "evaluatedOutputs": [
                {
                  "outputId": "Output_1",
                  "outputName": "",
                  "outputValue": "\"Jedi\""
                }
              ]
            }
          ],
          "decisionId": "jedi_or_sith",
          "decisionName": "Jedi or Sith",
          "decisionKey": 2251799813685265
        },
        {
          "decisionType": "DECISION_TABLE",
          "evaluatedInputs": [
            {
              "inputId": "InputClause_0qnqj25",
              "inputName": "",
              "inputValue": "\"Jedi\""
            },
            {
              "inputId": "InputClause_0k64hys",
              "inputName": "",
              "inputValue": "185"
            }
          ],
          "decisionVersion": 5,
          "decisionOutput": "\"Obi-Wan Kenobi\"",
          "matchedRules": [
            {
              "ruleId": "DecisionRule_0uin2hk",
              "ruleIndex": 2,
              "evaluatedOutputs": [
                {
                  "outputId": "OutputClause_0hhe1yo",
                  "outputName": "",
                  "outputValue": "\"Obi-Wan Kenobi\""
                }
              ]
            }
          ],
          "decisionId": "force_user",
          "decisionName": "Which force user?",
          "decisionKey": 2251799813685266
        }
      ],
      "evaluationFailureMessage": "",
      "failedDecisionId": "",
      "decisionRequirementsId": "force_users",
      "decisionRequirementsKey": 2251799813685264,
      "decisionId": "force_user",
      "decisionName": "Which force user?",
      "decisionKey": 2251799813685266,
      "bpmnProcessId": "force-user",
      "processDefinitionKey": 2251799813685268,
      "processInstanceKey": 2251799813685270,
      "elementInstanceKey": 2251799813685274,
      "elementId": "Activity_1sl5qf7"
    },
    "key": 2251799813685277,
    "timestamp": 1647429371592,
    "sourceRecordPosition": 46,
    "intent": "EVALUATED",
    "rejectionType": "NULL_VAL",
    "rejectionReason": "",
    "brokerVersion": "1.4.0",
    "valueType": "DECISION_EVALUATION",
    "recordType": "EVENT",
    "position": 50
  },
  "fields": {
    "value.decisionKey": [
      2251799813685266
    ],
    "value.evaluatedDecisions.decisionOutput": [
      "\"Jedi\"",
      "\"Obi-Wan Kenobi\""
    ],
    "value.evaluatedDecisions.decisionVersion": [
      4,
      5
    ],
    "sourceRecordPosition": [
      46
    ],
    "value.decisionRequirementsId": [
      "force_users"
    ],
    "value.evaluatedDecisions.decisionKey": [
      2251799813685265,
      2251799813685266
    ],
    "value.evaluatedDecisions.evaluatedInputs.inputName": [
      "",
      "",
      ""
    ],
    "rejectionType": [
      "NULL_VAL"
    ],
    "value.evaluatedDecisions.matchedRules.evaluatedOutputs.outputValue": [
      "\"Jedi\"",
      "\"Obi-Wan Kenobi\""
    ],
    "value.decisionVersion": [
      5
    ],
    "value.evaluatedDecisions.evaluatedInputs.inputId": [
      "Input_1",
      "InputClause_0qnqj25",
      "InputClause_0k64hys"
    ],
    "valueType": [
      "DECISION_EVALUATION"
    ],
    "value.decisionId": [
      "force_user"
    ],
    "value.elementId": [
      "Activity_1sl5qf7"
    ],
    "key": [
      2251799813685277
    ],
    "timestamp": [
      "2022-03-16T11:16:11.592Z"
    ],
    "value.processInstanceKey": [
      2251799813685270
    ],
    "value.evaluatedDecisions.evaluatedInputs.inputValue": [
      "\"blue\"",
      "\"Jedi\"",
      "185"
    ],
    "value.elementInstanceKey": [
      2251799813685274
    ],
    "value.decisionRequirementsKey": [
      2251799813685264
    ],
    "value.evaluatedDecisions.matchedRules.ruleId": [
      "DecisionRule_0zumznl",
      "DecisionRule_0uin2hk"
    ],
    "value.decisionOutput": [
      "\"Obi-Wan Kenobi\""
    ],
    "value.processDefinitionKey": [
      2251799813685268
    ],
    "partitionId": [
      1
    ],
    "value.evaluatedDecisions.matchedRules.evaluatedOutputs.outputName": [
      "",
      ""
    ],
    "recordType": [
      "EVENT"
    ],
    "value.decisionName": [
      "Which force user?"
    ],
    "value.failedDecisionId": [
      ""
    ],
    "brokerVersion": [
      "1.4.0"
    ],
    "intent": [
      "EVALUATED"
    ],
    "value.evaluatedDecisions.matchedRules.evaluatedOutputs.outputId": [
      "Output_1",
      "OutputClause_0hhe1yo"
    ],
    "value.bpmnProcessId": [
      "force-user"
    ],
    "value.evaluatedDecisions.decisionName": [
      "Jedi or Sith",
      "Which force user?"
    ],
    "value.evaluatedDecisions.matchedRules.ruleIndex": [
      1,
      2
    ],
    "value.evaluatedDecisions.decisionId": [
      "jedi_or_sith",
      "force_user"
    ],
    "value.evaluatedDecisions.decisionType": [
      "DECISION_TABLE",
      "DECISION_TABLE"
    ],
    "value.evaluationFailureMessage": [
      ""
    ],
    "position": [
      50
    ],
    "rejectionReason": [
      ""
    ]
  }
}
```

korthout commented:
Hiya @pihme, would you be willing to review this PR? If you want I can provide you with an intro to DMN in Zeebe. It's also fine if you'd rather someone else review. There's no need to rush this.

pihme reviewed (commented):
Saving comments; will continue after we had a chance to talk about it

pihme reviewed (approved):
LGTM :label:

korthout commented on engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/BusinessRuleTaskTest.java:
```diff
@@ -459,8 +461,146 @@ public void shouldWriteDecisionEvaluationEventIfEvaluationFailed() {
+    final var evaluatedDecisions = decisionEvaluationValue.getEvaluatedDecisions();
+    assertThat(evaluatedDecisions).hasSize(2);
+
+    assertThat(evaluatedDecisions.get(0))
```
👍 That's a good point. I did consider this initially but didn't know which approach would be better. I'll compact the test.

korthout commented on engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnDecisionBehavior.java:
```diff
@@ -311,12 +315,16 @@ private void addMatchedRuleToEvaluationEvent(
+
+    if (evaluatedOutput.outputName() != null) {
+      outputRecord.setOutputName(evaluatedOutput.outputName());
+    }
```
@pihme You pointed me to the Modeler's behavior that either shows a greyed out `Input` placeholder; or shows the input expression if an expression is provided and no input label/name is provided. This is pretty interesting behavior and makes me wonder what Operate displays.

@sdorokhova If a decision table's output does not have a name, what will Operate show? Likewise for a decision table's input. Does it use the input and output names in the UI?

korthout commented:
I'm postponing merging this until it is clear whether Operate needs Zeebe to provide input/output names in the same way Modeler shows them.

pihme commented on dmn/src/main/java/io/camunda/zeebe/dmn/impl/EvaluatedDmnScalaInput.java:
```diff
@@ -20,6 +21,16 @@ public static EvaluatedDmnScalaInput of(
+    if (input.name() != null) {
+      inputName = input.name();
+    } else if (input.expression() instanceof FeelExpression feelExpression) {
+      inputName = feelExpression.expression().text();
```
:wrench: Maybe cut string length to e.g. 30 characters

pihme reviewed (approved):
:tada: Looks good to me!

(One optional suggestion)

korthout commented:
bors merge

korthout commented:
bors merge

korthout commented:
bors merge

korthout commented:
Bors merge

korthout commented:
@pihme It seems there were some tests failing. I've adjusted the tests and updated the PR description to make it clear what this PR changes. Please have again another look 🙇

pihme reviewed (approved):
lgtm

korthout commented:
bors merge
